source/common/upstream/subset_lb.cc

```diff
@@ -0,0 +1,416 @@
+    return nullptr;
+  }
+
+  // Route has metadata match criteria defined, see if we have a matching subset.
+  LbSubsetEntryPtr entry = findSubset(match_criteria->metadataMatchCriteria());
+  if (entry == nullptr || !entry->active()) {
+    // No matching subset or subset not active: use fallback policy.
+    return nullptr;
+  }
+
+  host_chosen = true;
+  stats_.lb_subsets_selected_.inc();
+  return entry->lb_->chooseHost(context);
+}
+
+// Iterates over the given metadata match criteria (which must be
+// lexically sorted by key) and find a matching LbSubsetEnryPtr, if any.
+SubsetLoadBalancer::LbSubsetEntryPtr SubsetLoadBalancer::findSubset(
+    const std::vector<Router::MetadataMatchCriterionConstSharedPtr>& match_criteria) {
+  const LbSubsetMap* subsets = &subsets_;
+
+  // Because the match_criteria and the host metadata used to populate
+  // subsets_ are sorted in the same order, we can iterate over the
+  // criteria and perform a lookup for each key and value, starting
+  // with the root LbSubsetMap and using the previous iteration's
+  // LbSubsetMap thereafter (tracked in subsets). If ever a
+  // criterion's key or value is not found, there is no subset for
+  // this criteria. If we reach the last criterion, we've found the
+  // LbSubsetEntry for the criteria, which may or may not have a
+  // subset attached to it.
+  for (uint32_t i = 0; i < match_criteria.size(); i++) {
+    const Router::MetadataMatchCriterion& match_criterion = *match_criteria[i];
+    const auto& subset_it = subsets->find(match_criterion.name());
+    if (subset_it == subsets->end()) {
+      // No subsets with this key (at this level in the hierarchy).
+      break;
+    }
+
+    const ValueSubsetMap& vs_map = subset_it->second;
+    const auto& vs_it = vs_map.find(match_criterion.value());
+    if (vs_it == vs_map.end()) {
+      // No subsets with this value.
+      break;
+    }
+
+    const LbSubsetEntryPtr& entry = vs_it->second;
+    if (i + 1 == match_criteria.size()) {
+      // We've reached the end of the criteria, and they all
+      // matched.
+      return entry;
+    }
+
+    subsets = &entry->children_;
+  }
+
+  return nullptr;
+}
+
+// Given the addition and/or removal of hosts, update all subsets,
+// creating new subsets as necessary.
+void SubsetLoadBalancer::update(const std::vector<HostSharedPtr>& hosts_added,
+                                const std::vector<HostSharedPtr>& hosts_removed) {
+  if (fallback_policy_ == envoy::api::v2::Cluster::LbSubsetConfig::ANY_ENDPOINT) {
+    HostPredicate predicate = [](const Host&) -> bool { return true; };
+    if (fallback_subset_ == nullptr) {
+      // First update: create the default host subset with all hosts
+      fallback_subset_.reset(new LbSubsetEntry());
+      fallback_subset_->initLoadBalancer(*this, predicate);
+    } else {
+      // Subsequent updates: add/remove hosts
+      fallback_subset_->host_subset_->update(hosts_added, hosts_removed, predicate);
```

Comment:
Nit: Given duplication here and at 150, is it possible to make the logic flow so that we always call update() on the fallback_subset_ and just have an initialize step if it doesn't yet exist? That might make the logic easier to quickly grok. If this is a big deal feel free to ignore.